syntax = "proto3";

package magma.lte;

// Proto file to serialize the structures in
// magma/lte/gateway/c/oai/common/common_types.h

// guti_t
message Guti {
  bytes plmn = 1;
  uint32 mme_gid = 2;
  uint32 mme_code = 3;
  uint32 m_tmsi = 4;
}

// ecgi_t
message Ecgi {
  bytes plmn = 1; //char[] as plmn_t is 3 bytes
  // serializing struct eci_t here without creating a new message
  uint32 enb_id = 2;
  uint32 cell_id = 3;
  uint32 empty = 4;
}

// Messages for APN configuration

// ip_address_t
message IpAddr {
  uint32 pdn_type = 1; // pdn_type_value_t
  uint64 ipv4_address = 2; // in_addr
  string ipv6_address = 3; // in6_addr
}

// eps_subscribed_qos_profile_t
message EpsSubscribedQosProfile {
  uint32 qci = 1; // qci_t
  // serializing allocation_retention_priority_t here instead of adding a
  // new message type
  uint32 priority_level = 2; // priority_level_t
  uint32 pre_emption_capability = 3; // pre_emption_capability_t
  uint32 pre_emption_vulnerability = 4; // pre_emption_vulnerability_t
}

// ambr_t
message Ambr {
  uint64 br_ul = 1;
  uint64 br_dl = 2;
}

// apn_configuration_t
message ApnConfig {
  uint32 context_identifier = 1;
  uint32 num_ip_addrs = 2;
  repeated IpAddr ip_address = 3;
  uint32 pdn_type = 4;
  string service_selection = 5;
  uint32 service_selection_len = 6;
  EpsSubscribedQosProfile subscribed_qos = 7;
  Ambr ambr = 8;
}

// apn_config_profile
message ApnConfigProfile {
  uint32 context_identifier = 1;
  uint32 all_apn_conf_ind = 2; // all_apn_conf_ind_t
  uint32 num_apns = 3;
  repeated ApnConfig apn_configs = 4;
}
